web/src/engine/predictive-text/worker-thread/docs/search-spaces.md

```diff
@@ -0,0 +1,76 @@
+    - How much of the context should be altered when applying these suggestions?
+<!--   - To complicate matters further, note that the letters `c`, `v`, and `n` are also close to `b`.
+    - Suppose this leads to `van errors`, `NaN errors`, etc..., but also `cannery`, `Vannessa`, etc. -->
+
+2.  Each individual `SearchSpace` should only model correction of inputs that result in tokens of the same codepoint length as each other.
+    - Consider the effects of a long left-deletion input.
+        - If a `SearchSpace` held variable-length text, a large left-deletion could delete the whole block of text for some cases, but not for others; this gets messy to handle!
+    - We want to ensure that `deleteLeft` effects can be modeled clearly and consistently, without ambiguity.
+
+### Analysis
+
+1.  Each keystroke's probability may be considered independent from other keystrokes.
+    - To be clear, this is a simplifying assumption; attempting to do otherwise would both be speculative and complicate things further.
+
+2.  It is not possible to guarantee that one keystroke will only extend a previous `SearchSpace` in one way.
+    - If the incoming keystroke produces `Transform`s that have different `insert` length without varying the left-deletion count, this _must_ result in multiple `SearchSpace`s, as the total codepoint length will vary accordingly.
+    - Also of note:  if left-deleting, it is possible for a left-deletion to erase the token adjacent to the text insertion point.
+
+3.  When constructing and applying `Suggestion`s, it helps greatly to determine which `SearchSpace` led to it.
+    - This allows us to determine _which_ keystrokes are being replaced, as well as _what_ parts of the Context will be affected.
+    - If a correction is only possible if the context were to have a different initial word-boundary than currently exists, this implies a need to correct the Context accordingly as part of applying a suggestion.
+
+4.  Following from the points above, it is possible for two or more `SearchSpace` segments to model input sequences with the same properties, as follows:
+    - The input length in codepoints is identical.
+    - The same keystrokes, and portions thereof, are utilized.
+
+For example, consider a case with two keystrokes, each of which has versions emitting insert strings of one and two characters.  Taking two chars from one and one char from the other will result in a `SearchSpace` that models a total of two keystrokes that fully covers the two keys.
+
+For such cases, any future keystrokes can extend both input sequences in the same manner.  While the actual correction-text may differ, the net effect it has on the properties of a token necessary for correction and construction of suggestions is identical.  The `SearchCluster` variant of `SearchSpace` exists for such cases, modeling the convergence of multiple `SearchPath`s and extending all of them together at once.
```

Comment:
This example doesn't make sense to me. I don't understand "each of which has versions emitting insert strings of one and two characters."